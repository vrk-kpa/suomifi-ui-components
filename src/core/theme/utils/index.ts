export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
