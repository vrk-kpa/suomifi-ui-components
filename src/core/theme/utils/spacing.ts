import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (props: SpacingProps | undefined): string => {
  if (!props) return '';

  const cssStyles = Object.entries(props)
    .map(([key, value]) =>
      getCSSSpacing(defaultSuomifiTheme, key as keyof SpacingProps, value),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return `margin-right: ${amount}; margin-left: ${amount};`;
    case 'my':
      return `margin-top: ${amount}; margin-bottom: ${amount};`;
    case 'px':
      return `padding-right: ${amount}; padding-left: ${amount};`;
    case 'py':
      return `padding-top: ${amount}; padding-bottom: ${amount};`;
    default:
      return `${[cssSelector[key]]}: ${amount};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
