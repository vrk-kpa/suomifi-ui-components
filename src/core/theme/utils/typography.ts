import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
