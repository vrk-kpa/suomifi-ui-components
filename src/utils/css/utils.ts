import { css } from 'styled-components';
import { cssValueToString } from './cssvalue';

const camelToSnake = (string: string) =>
  string.replace(/[\w]([A-Z])/g, (m) => `${m[0]}-${m[1]}`).toLowerCase();
/**
 * Convert CSSObject to CSS FlattenSimpleInterpolation
 * @param value CSSObject
 *
 * @example
 *  cssObjectToCss({
 *    fontFamily: "'Arial', sans-serif";
 *    fontSize: {value: 16, unit: 'px'};
 *  })
 */
export const cssObjectToCss = <T extends Record<string, any>>(object: T) => css`
  ${Object.entries(object)
    .map(([key, value]) => `${camelToSnake(key)}: ${cssValueToString(value)};`)
    .join('')}
`;

/**
 * Convert CSSObject tokens to CSS FlattenSimpleInterpolation
 * @param tokens Tokens of CSSObjects
 *
 * @example
 *  cssObjectsToCss({
 *    bodyText: {
 *      fontFamily: "'Arial', sans-serif";
 *      fontSize: {value: 16, unit: 'px'};
 *    },
 *    bodyTextSmallScreen: {
 *      fontFamily: "'Arial', sans-serif";
 *      fontSize: {value: 18, unit: 'px'};
 *     }
 *  })
 */
export const cssObjectsToCss = <
  T extends Record<string, any>,
  K extends keyof T,
>(
  tokens: T,
) =>
  Object.entries(tokens).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: css`
        ${cssObjectToCss(value)}
      `,
    }),
    {} as { [key in K]: ReturnType<typeof css> },
  );
