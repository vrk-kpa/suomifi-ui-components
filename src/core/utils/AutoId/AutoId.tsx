/** Based heavily on https://github.com/reach/reach-ui/tree/develop/packages/auto-id */
/* As a11y API's needs element IDs and requiring user to give those all not suitable, we generate them automatically.
 * Generating random IDs are not working as those will not match with the client & server side rendering.
 *
 * Idea of the AutoId is to patch incremented id after the first render.
 */
import { useState, useEffect } from 'react';
import { useEnhancedEffect } from '../../../utils/common';
import { useConfig } from '../ConfigProvider/ConfigProvider';

let autoId = 0;
let clientRender = false;

const generateNextId = (prefix?: string) => {
  autoId += 1;
  console.log('idPrefix', prefix);
  if (prefix) {
    return `${prefix}-${autoId}`;
  }
  return autoId;
};

const useGeneratedId = (propId?: string | null, idPrefix?: string) => {
  const startId = propId || (clientRender ? generateNextId() : null);
  const [generatedId, setId] = useState(startId);

  useEnhancedEffect(() => {
    if (generatedId === null) {
      setId(generateNextId(idPrefix));
    }
  }, []);

  useEffect(() => {
    if (!clientRender) {
      clientRender = true;
    }
  }, []);
  return generatedId != null ? String(generatedId) : undefined;
};

/**
 * Returns automatically generated id if one was not provided.
 */
export const AutoId = ({
  id,
  children,
}: {
  id?: string;
  children: (passedId: string) => JSX.Element;
}) => {
  const { idPrefix } = useConfig();
  const generatedId = useGeneratedId(id, idPrefix);
  return children(!!generatedId ? generatedId : '');
};
