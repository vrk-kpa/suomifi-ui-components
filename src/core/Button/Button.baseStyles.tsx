import { css } from 'styled-components';
import { SuomifiTheme } from '../theme';
import { button } from '../theme/reset';
import { alphaHex } from '../../utils/css';
import { MarginProps, buildSpacingCSS } from '../theme/utils/spacing';

const invertedStyles = (theme: SuomifiTheme) => css`
  &.fi-button--inverted {
    background: none;
    background-color: ${theme.colors.highlightBase};
    border: 1px solid ${theme.colors.whiteBase};
    text-shadow: none;

    &:hover {
      background: ${theme.gradients.whiteBaseNegative};
    }

    &:active {
      background: none;
      background-color: ${theme.colors.highlightBase};
    }

    &.fi-button--disabled,
    &[disabled],
    &:disabled {
      opacity: 0.41;
      background: none;
      background-color: none;
    }
  }
`;

const secondary = (theme: SuomifiTheme) => css`
  color: ${theme.colors.highlightBase};
  background: none;
  background-color: ${theme.colors.whiteBase};
  border: 1px solid ${theme.colors.highlightBase};
  text-shadow: none;

  &:hover {
    background: ${theme.gradients.whiteBaseToDepthLight2};
  }

  &:active {
    background: none;
    background-color: ${theme.colors.depthLight2};
  }

  &.fi-button--disabled,
  &[disabled],
  &:disabled {
    color: ${theme.colors.depthBase};
    text-shadow: none;
    background: none;
    background-color: ${theme.colors.highlightLight4};
    border-color: ${theme.colors.depthBase};
  }

  .fi-button_loading-icon {
    .fi-icon-component-brand-fill {
      fill: ${theme.colors.depthBase};
    }
  }
`;

const secondaryStyles = (theme: SuomifiTheme) => css`
  &.fi-button--secondary {
    ${secondary(theme)}
  }
`;

const secondaryNoBorderStyles = (theme: SuomifiTheme) => css`
  &.fi-button--secondary-noborder {
    ${secondary(theme)}
    border: none;
    padding: ${theme.spacing.insetL} ${theme.spacing.insetXxl};
    background-color: transparent;

    &.fi-button--icon-only {
      padding: ${theme.spacing.insetS} 12px;
    }

    .fi-button_loading-icon {
      .fi-icon-component-brand-fill {
        fill: ${theme.colors.depthBase};
      }
    }
  }
`;

const secondaryLightStyles = (theme: SuomifiTheme) => css`
  &.fi-button--secondary-light {
    color: ${theme.colors.highlightBase};
    ${secondary(theme)}
    background: ${theme.gradients.depthSecondaryToDepthSecondaryDark1};
    padding: ${theme.spacing.insetL} ${theme.spacing.insetXxl};
    border: none;

    &.fi-button--icon-only {
      padding: ${theme.spacing.insetS} 12px;
    }

    &:hover {
      background: ${theme.gradients.highlightLight4ToDepthSecondary};
    }

    &:active {
      background: ${theme.gradients.depthLight3ToDepthLight2};
    }

    &.fi-button--disabled,
    &[disabled],
    &:disabled {
      color: ${theme.colors.depthBase};
      background: none;
      background-color: ${theme.colors.depthLight3};
    }

    .fi-button_loading-icon {
      .fi-icon-component-brand-fill {
        fill: ${theme.colors.depthBase};
      }
    }
  }
`;

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${button(theme)}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  padding: 9px ${theme.spacing.insetXxl};
  min-height: 40px;
  color: ${theme.colors.whiteBase};
  background: ${theme.gradients.highlightBaseToHighlightDark1};
  border-radius: ${theme.radiuses.basic};
  text-align: center;
  text-shadow: ${theme.shadows.invertTextShadow};
  cursor: pointer;
  border: 1px solid transparent; /* For high contrast mode */

  &:focus {
    position: relative;
    ${theme.focuses.highContrastFocus} /* For high contrast mode */

    &::after {
      ${theme.focuses.absoluteFocus}
    }
  }

  &:hover {
    background: ${theme.gradients.highlightLight1ToHighlightBase};
    outline: 2px solid transparent; /* For high contrast mode */
  }

  &:active {
    background: ${theme.colors.highlightDark1};
  }

  &.fi-button--disabled,
  &[disabled],
  &:disabled {
    text-shadow: 0 1px 1px ${alphaHex(0.5)(theme.colors.blackBase)};
    background: ${theme.gradients.depthLight1ToDepthBase};
    user-select: none;
    cursor: not-allowed;
  }

  &.fi-button--disabled::after {
    border: none;
    box-shadow: none;
  }

  &.fi-button--fullwidth {
    display: block;
    width: 100%;
  }

  ${invertedStyles(theme)}
  ${secondaryStyles(theme)}
  ${secondaryNoBorderStyles(theme)}

  ${secondaryLightStyles(theme)}
  
  & > .fi-button_icon > .fi-icon {
    width: 16px;
    height: 16px;
    margin-right: ${theme.spacing.insetM};
    vertical-align: middle;
    transform: translateY(-0.1em);
  }

  & > .fi-button_icon--right > .fi-icon {
    margin-right: 0;
    margin-left: ${theme.spacing.insetM};
  }

  &.fi-button--icon-only {
    padding: ${theme.spacing.insetS} 11px;
    & > .fi-button_icon > .fi-icon {
      margin-right: 0;
    }
    & > .fi-button_icon--right > .fi-icon {
      margin-left: 0;
    }
  }

  &.fi-button--disabled > .fi-button_icon {
    cursor: not-allowed;
  }

  .fi-button_loading-icon {
    display: inline-block;
    width: 20px;
    height: 20px;
    margin-right: ${theme.spacing.insetM};
    animation: rotation 1.5s infinite linear;

    .fi-icon-component-brand-fill {
      fill: ${theme.colors.whiteBase};
    }
  }

  @keyframes rotation {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(359deg);
    }
  }
  @media (prefers-reduced-motion) {
    &.fi-loadingSpinner.fi-loadingSpinner--loading {
      .fi-loadingSpinner_icon {
        animation: rotation 10s infinite linear;
      }
    }
  }

  &.fi-button--loading {
    display: flex;
    align-items: center;
  }

  &.fi-button--loading-icon-only {
    .fi-button_loading-icon {
      margin-right: 0;
    }
  }
`;
