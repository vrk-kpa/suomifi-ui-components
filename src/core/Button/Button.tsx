import React, { Component, forwardRef, ReactElement, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { baseStyles } from './Button.baseStyles';
import { HtmlButton, HtmlButtonProps, HtmlSpan } from '../../reset';
import { SuomifiThemeConsumer, SuomifiThemeProp } from '../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMargins,
} from '../theme/utils/spacing';
import { IconPreloader } from 'suomifi-icons';
import { VisuallyHidden } from '../VisuallyHidden/VisuallyHidden';
import { SpacingConsumer } from '../theme/SpacingProvider/SpacingProvider';
import { filterDuplicateKeys } from '../../utils/common/common';

export type ButtonVariant =
  | 'default'
  | 'inverted'
  | 'secondary'
  | 'secondaryNoBorder'
  | 'secondaryLight';

export type ForcedAccessibleNameProps =
  | {
      children: ReactNode;
      'aria-label'?: string;
    }
  | {
      /**
       * Button element content
       */
      children?: never;
      /**
       * Define a label if button's text content does not indicate the button's purpose (for example, button with only an icon).
       * If the button has a visible label, make sure the aria-label includes the visible text.
       * Alternatively you can define an `aria-labelledby`.
       */
      'aria-label': string;
    };

export type LoadingProps =
  | {
      loading?: false;
      ariaLoadingText?: never;
    }
  | {
      /** Shows the animated icon indicating a loading state
       * @default false
       */
      loading?: true;
      /** Text for assistive technology to indicate loading state. Required with `loading` */
      ariaLoadingText: string;
    };

interface InternalButtonProps
  extends Omit<HtmlButtonProps, 'aria-disabled' | 'onClick'>,
    MarginProps {
  /**
   * Variant for the button
   * `default` | `inverted` | `secondary` | `secondaryNoBorder` | `secondaryLight`
   * @default default
   */
  variant?: ButtonVariant;
  /** Disables the button */
  disabled?: boolean;
  /** Soft disables the button to allow tab-focus. Disables onClick() functionality */
  'aria-disabled'?: boolean;
  /** CSS class for custom styles */
  className?: string;
  /**
   * Sets width to fill all available space
   */
  fullWidth?: boolean;
  /**
   * Icon from suomifi-icons
   */
  icon?: ReactElement;
  /**
   * Icon from suomifi-icons to be placed on the right side
   */
  iconRight?: ReactElement;
  /** Callback fired on button click */
  onClick?: (event: React.MouseEvent) => void;
  /** Ref object is passed to the button element. Alternative to React `ref` attribute. */
  forwardedRef?: React.RefObject<HTMLButtonElement>;
}

export type ButtonProps = InternalButtonProps &
  ForcedAccessibleNameProps &
  LoadingProps & {
    globalMargins?: GlobalMargins;
  };

const baseClassName = 'fi-button';
const disabledClassName = `${baseClassName}--disabled`;
const iconClassName = `${baseClassName}_icon`;
const iconStandaloneClassName = `${baseClassName}--icon-only`;
const iconRightClassName = `${baseClassName}_icon--right`;
const fullWidthClassName = `${baseClassName}--fullwidth`;
const loadingButtonClassName = `${baseClassName}--loading`;
const loadingButtonIconOnlyClassName = `${baseClassName}--loading-icon-only`;
const loadingIconClassName = `${baseClassName}_loading-icon`;

class BaseButton extends Component<ButtonProps> {
  render() {
    const {
      fullWidth,
      variant = 'default',
      className,
      loading,
      ariaLoadingText,
      disabled,
      onClick,
      'aria-disabled': ariaDisabled = loading || false,
      icon,
      iconRight,
      forwardedRef,
      children,
      style,
      ...rest
    } = this.props;
    const [_marginProps, passProps] = separateMarginProps(rest);
    const onClickProp = !!disabled || !!ariaDisabled ? null : { onClick };

    return (
      <>
        <VisuallyHidden aria-live="polite">
          {loading && ariaLoadingText}
        </VisuallyHidden>
        <HtmlButton
          {...passProps}
          {...onClickProp}
          {...(!!disabled ? {} : { tabIndex: 0 })}
          aria-disabled={!!ariaDisabled || !!disabled}
          forwardedRef={forwardedRef}
          disabled={!!disabled}
          className={classnames(baseClassName, className, {
            [disabledClassName]: !!disabled || !!ariaDisabled,
            [`${baseClassName}--inverted`]: variant === 'inverted',
            [`${baseClassName}--secondary`]: variant === 'secondary',
            [`${baseClassName}--secondary-noborder`]:
              variant === 'secondaryNoBorder',
            [`${baseClassName}--secondary-light`]: variant === 'secondaryLight',
            [fullWidthClassName]: fullWidth,
            [iconStandaloneClassName]: (!!icon || !!iconRight) && !children,
            [loadingButtonClassName]: loading,
            [loadingButtonIconOnlyClassName]: loading && !children,
          })}
          style={style}
        >
          {loading && <IconPreloader className={loadingIconClassName} />}
          {!!icon && !loading && (
            <HtmlSpan className={iconClassName}>{icon}</HtmlSpan>
          )}
          {children}
          {!!iconRight && !loading && (
            <HtmlSpan className={classnames(iconClassName, iconRightClassName)}>
              {iconRight}
            </HtmlSpan>
          )}
        </HtmlButton>
      </>
    );
  }
}

const StyledButton = styled(
  ({
    theme,
    globalMargins,
    ...passProps
  }: ButtonProps & SuomifiThemeProp & { globalMargins: GlobalMargins }) => (
    <BaseButton {...passProps} />
  ),
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.button,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const Button = forwardRef(
  (props: ButtonProps, ref: React.RefObject<HTMLButtonElement>) => (
    <SpacingConsumer>
      {({ margins }) => (
        <SuomifiThemeConsumer>
          {({ suomifiTheme }) => (
            <StyledButton
              theme={suomifiTheme}
              forwardedRef={ref}
              globalMargins={margins}
              {...props}
            />
          )}
        </SuomifiThemeConsumer>
      )}
    </SpacingConsumer>
  ),
);

export const InternalButton = forwardRef(
  (props: ButtonProps, ref: React.RefObject<HTMLButtonElement>) => (
    <SuomifiThemeConsumer>
      {({ suomifiTheme }) => (
        <StyledButton theme={suomifiTheme} forwardedRef={ref} {...props} />
      )}
    </SuomifiThemeConsumer>
  ),
);

Button.displayName = 'Button';
export { Button };
