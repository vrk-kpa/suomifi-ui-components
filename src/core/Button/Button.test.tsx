import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../utils/test';
import { AutoId } from '../utils/AutoId/AutoId';

import { Button } from './Button';

describe('Basic Button', () => {
  it('render with the same component on the same container does not remount', () => {
    const { getByRole, rerender } = render(
      <AutoId>{(id) => <Button id={id}>Test</Button>}</AutoId>,
    );
    const button = getByRole('button');
    expect(button.textContent).toBe('Test');
    expect(button).toHaveAttribute('id', '1');
    // re-render the same component with different props
    rerender(<AutoId>{(id) => <Button id={id}>Test two</Button>}</AutoId>);
    const rerendered = getByRole('button');
    expect(rerendered.textContent).toBe('Test two');
    expect(rerendered).toHaveAttribute('id', '1');
  });

  it(
    'should not have basic accessibility issues',
    axeTest(<Button>Test</Button>),
  );
});

describe('Disabled Button', () => {
  it('is disabled and changes to enabled', () => {
    const { getByRole, rerender } = render(
      <Button disabled={true}>Test</Button>,
    );
    expect(getByRole('button')).toBeDisabled();

    rerender(<Button disabled={false}>Test</Button>);
    expect(getByRole('button')).toBeEnabled();
  });

  it('is aria-disabled and changes to enabled', () => {
    const { getByRole, rerender } = render(
      <Button aria-disabled={true}>Aria-disabled button</Button>,
    );
    expect(getByRole('button')).toHaveAttribute('aria-disabled', 'true');
    rerender(<Button aria-disabled={false}>Aria-disabled button</Button>);
    expect(getByRole('button')).toHaveAttribute('aria-disabled', 'false');
  });
});

describe('Button variant', () => {
  it('default should match snapshot', () => {
    const { container } = render(<Button>Button</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('inverted should match snapshot', () => {
    const { container } = render(
      <Button variant="inverted">Inverted button</Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('secondary should match snapshot', () => {
    const { container } = render(
      <Button variant="secondary">Secondary button</Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('secondaryNoBorder should match snapshot', () => {
    const { container } = render(
      <Button variant="secondaryNoBorder">secondaryNoBorder button</Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('link match snapshot', () => {
    const { container } = render(
      <Button variant="secondaryLight">Secondary light button</Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('Margin prop', () => {
    it('should have margin style from margin prop', () => {
      const { container } = render(<Button margin="xs">Test button</Button>);
      const button = container.querySelector('button');
      expect(button).toHaveStyle('margin: 10px');
    });

    it('should have margin prop style overwritten from style', () => {
      const { container } = render(
        <Button margin="xs" style={{ margin: 2 }}>
          Test button
        </Button>,
      );
      const button = container.querySelector('button');
      expect(button).toHaveAttribute('style', 'margin: 2px;');
    });
  });

  describe('Loading spinner', () => {
    it('should display a loading spinner when loading prop is used', () => {
      const { container } = render(
        <Button loading ariaLoadingText="Loading">
          Test button
        </Button>,
      );
      const spinner = container.querySelector('svg');
      expect(spinner).toBeInTheDocument();
    });
    it('should render a visually hidden aria-live text when loading prop is used', async () => {
      const { container, rerender } = render(<Button>Test button</Button>);
      const visuallyHiddenElement = container.querySelector(
        '.fi-visually-hidden',
      );
      expect(visuallyHiddenElement).toBeEmptyDOMElement();

      rerender(
        <Button loading ariaLoadingText="Loading">
          Test button
        </Button>,
      );
      const visuallyHiddenElementWithLoadingState = container.querySelector(
        '.fi-visually-hidden',
      );
      expect(visuallyHiddenElementWithLoadingState).toHaveTextContent(
        'Loading',
      );
    });
  });
});
