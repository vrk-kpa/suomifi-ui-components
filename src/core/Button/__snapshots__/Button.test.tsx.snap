// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button variant default should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 ::before,
.c0 ::after {
  box-sizing: border-box;
}

.c1 {
  position: absolute;
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

@media (prefers-reduced-motion) {

}

<span
  aria-live="polite"
  class="c0 c1 fi-visually-hidden"
/>
`;

exports[`Button variant inverted should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 ::before,
.c0 ::after {
  box-sizing: border-box;
}

.c1 {
  position: absolute;
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

@media (prefers-reduced-motion) {

}

<span
  aria-live="polite"
  class="c0 c1 fi-visually-hidden"
/>
`;

exports[`Button variant link match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 ::before,
.c0 ::after {
  box-sizing: border-box;
}

.c1 {
  position: absolute;
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

@media (prefers-reduced-motion) {

}

<span
  aria-live="polite"
  class="c0 c1 fi-visually-hidden"
/>
`;

exports[`Button variant secondary should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 ::before,
.c0 ::after {
  box-sizing: border-box;
}

.c1 {
  position: absolute;
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

@media (prefers-reduced-motion) {

}

<span
  aria-live="polite"
  class="c0 c1 fi-visually-hidden"
/>
`;

exports[`Button variant secondaryNoBorder should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 ::before,
.c0 ::after {
  box-sizing: border-box;
}

.c1 {
  position: absolute;
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

@media (prefers-reduced-motion) {

}

<span
  aria-live="polite"
  class="c0 c1 fi-visually-hidden"
/>
`;
